biz.aQute.bndlib/src/aQute/bnd/classfile/ConstantPool.java

```diff
@@ -888,4 +898,316 @@ public void write(DataOutput out) throws IOException {
 			}
 		}
 	}
+
+	/**
+	 * A visitor of each entry with the proper types. Saves a lot of
+	 * 'instanceof' checking. See #
+	 */
+
+	@ProviderType
+	public interface EntryVisitor {
```

Comment:
I tried very hard to avoid visitor patterns since it is the main problem I have with asm. If you need this, cannot it not be done in a subclass of ConstantPool in a subpackage of classfile?

Reply:
You don't need to use instanceof in nested ifs. Do `switch (contant_pool.tag(index))` and the cases are the `CONSTANT_` values. See https://github.com/bndtools/bnd/blob/0eca95e0e2c99bfd22d495fa368d355eed624d1c/biz.aQute.bndlib/src/aQute/bnd/osgi/Clazz.java#L955-L957

Similarly for attributes. Do `switch (attribute.name())` and the cases are the attribute NAMEs. See https://github.com/bndtools/bnd/blob/0eca95e0e2c99bfd22d495fa368d355eed624d1c/biz.aQute.bndlib/src/aQute/bnd/osgi/Clazz.java#L976-L981

Both of these avoid `instanceof` tests and also avoid the future compatibility problems of visitor interfaces which need extending whenever a new thing is added. See asm :-(